### ISSUE TITLE
PageView Builder is creating upcoming page asymetrically.

### ISSUE THREAD

**Issue by jigarpatel17** [flutter/flutter#60295]
### Details
I have mentioned sample below. In this case, we can see that for page index 3, item builder is calling at the time of creating page for index 2. Assuming that PageView is trying to create next page in advance which seems like good feature. But believing that this behavior should be symmetric throughout all the pages.

### Code snippet
```
final PageController pageController = PageController();
  
@override
Widget build(BuildContext context) {
  return Container(
      child: PageView.builder(
          itemCount: 10,
          onPageChanged: _handlePageChanged,
          controller: pageController,
          scrollDirection: Axis.vertical,
          itemBuilder: (context, position) {
            print('itemBuilder position: $position');
            return Container(
              color: RandomColor().randomColor(),
            );
          }
        )
    );
}

void _handlePageChanged(int page) {
  print('_handlePageChanged: $page');
}
```

### Console logs
```
2020-06-25 22:46:41.669 I: itemBuilder position: 0                                                                              
2020-06-25 22:47:24.582 W: mapper 3.x is not supported
2020-06-25 22:47:58.013 I: itemBuilder position: 1
2020-06-25 22:47:58.062 I: _handlePageChanged: 1
2020-06-25 22:48:03.607 I: itemBuilder position: 2
2020-06-25 22:48:03.808 I: _handlePageChanged: 2
2020-06-25 22:48:04.595 I: itemBuilder position: 3
2020-06-25 22:48:09.468 I: _handlePageChanged: 3
2020-06-25 22:48:12.291 I: itemBuilder position: 4
2020-06-25 22:48:12.523 I: _handlePageChanged: 4
2020-06-25 22:48:13.759 I: itemBuilder position: 5
2020-06-25 22:48:13.927 I: _handlePageChanged: 5
2020-06-25 22:48:17.953 I: itemBuilder position: 6
2020-06-25 22:48:18.135 I: _handlePageChanged: 6
2020-06-25 22:48:20.189 I: itemBuilder position: 7
2020-06-25 22:48:20.406 I: _handlePageChanged: 7
2020-06-25 22:48:26.352 I: itemBuilder position: 8
2020-06-25 22:48:26.517 I: _handlePageChanged: 8
2020-06-25 22:48:27.820 I: itemBuilder position: 9
2020-06-25 22:48:28.004 I: _handlePageChanged: 9
```

Here, item builder for 3rd index is called at 2020-06-25 22:48:04.595 which is immediate after 2nd index page. The same thing is not happening for any other index and the same behavior, able to reproduce it every time.

### Expected behavior:
- PageView should provide param through which we can define page count which it can auto-create in advance to make user experience smooth.
- Auto creation of page should be symmetric for all page index. May be we can have bool param to define the same.

### Flutter and device details
```
jigar$ flutter doctor -v 
[✓] Flutter (Channel stable, v1.17.4, on Mac OS X 10.15.5 19F101, locale en-IN)
    • Flutter version 1.17.4 at /Users/jigar/Desktop/Work/sdk/flutter
    • Framework revision 1ad9baa8b9 (8 days ago), 2020-06-17 14:41:16 -0700
    • Engine revision ee76268252
    • Dart version 2.8.4

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0-rc2)
    • Android SDK at /Users/jigar/Library/Android/sdk
    • Platform android-30, build-tools 30.0.0-rc2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3.1, Build version 11C504
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] Connected device (1 available)
    • Pixel 3a • xxxxxxxxx • android-arm64 • Android 10 (API 29)

• No issues found!
```

Thank you for the time in advance

</details>

**Comment by VladyslavBondarenko** [comment]
Hi @jigarpatel17 
are you swiping `1` hard in such way that the page `3` is shown for a moment after `2`?
(the next page is built in this case)

**Comment by jigarpatel17** [comment]
@VladyslavBondarenko I have tried to disable scroll using **_NeverScrollableScrollPhysics_** and added button to move to next screen, still getting same issue. Also my all pages are having full screen height.

**Comment by VladyslavBondarenko** [comment]
I see this with always with page 7 and only with page 7: it's built right after `_handlePageChanged: 6`, and not right before `_handlePageChanged: 7`:

Log
```
I/flutter (31946): itemBuilder position: 0, time: 2020-06-29 13:46:12.402193
I/flutter (31946): itemBuilder position: 1, time: 2020-06-29 13:46:16.126998
I/flutter (31946): _handlePageChanged: 1, time: 2020-06-29 13:46:16.149514
I/flutter (31946): itemBuilder position: 2, time: 2020-06-29 13:46:20.159518
I/flutter (31946): _handlePageChanged: 2, time: 2020-06-29 13:46:20.204512
I/flutter (31946): itemBuilder position: 3, time: 2020-06-29 13:46:23.758252
I/flutter (31946): _handlePageChanged: 3, time: 2020-06-29 13:46:23.819283
I/flutter (31946): itemBuilder position: 4, time: 2020-06-29 13:46:28.059962
I/flutter (31946): _handlePageChanged: 4, time: 2020-06-29 13:46:28.120876
I/flutter (31946): itemBuilder position: 5, time: 2020-06-29 13:46:31.624571
I/flutter (31946): _handlePageChanged: 5, time: 2020-06-29 13:46:31.654699
I/flutter (31946): itemBuilder position: 6, time: 2020-06-29 13:46:34.844488
I/flutter (31946): _handlePageChanged: 6, time: 2020-06-29 13:46:34.907703
I/flutter (31946): itemBuilder position: 7, time: 2020-06-29 13:46:34.919218
I/flutter (31946): _handlePageChanged: 7, time: 2020-06-29 13:46:39.358776
I/flutter (31946): itemBuilder position: 8, time: 2020-06-29 13:46:43.663400
I/flutter (31946): _handlePageChanged: 8, time: 2020-06-29 13:46:43.709521
I/flutter (31946): itemBuilder position: 9, time: 2020-06-29 13:46:47.665075
I/flutter (31946): _handlePageChanged: 9, time: 2020-06-29 13:46:47.726884
```

Code sample

```
import 'package:flutter/material.dart';

void main() => runApp(MaterialApp(home: MyApp()));

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final PageController pageController = PageController();
  List colors = List.generate(10, (index) => index.isEven ? Colors.red : Colors.blue);

  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: Stack(
        children: [
          PageView.builder(
              itemCount: 10,
              onPageChanged: _handlePageChanged,
              controller: pageController,
              scrollDirection: Axis.vertical,
              physics: NeverScrollableScrollPhysics(),
              itemBuilder: (context, position) {
                print('itemBuilder position: $position, time: ${DateTime.now()}');
                return Container(color: colors[position]);
              }),
          Positioned(
            child: RaisedButton(
              child: Text('next page'),
              onPressed: () async {
                await pageController.nextPage(duration: Duration(milliseconds: 100), curve: Curves.bounceIn);
              },
            ),
          )
        ],
      ),
    );
  }

  void _handlePageChanged(int page) {
    print('_handlePageChanged: $page, time: ${DateTime.now()}');
  }
}
```

<details>
 <summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 1.20.0-3.0.pre.78, on Mac OS X 10.15.5 19F101, locale en-GB)
    • Flutter version 1.20.0-3.0.pre.78 at /Users/nevercode/dev/flutter
    • Framework revision 2962912d68 (2 days ago), 2020-06-27 00:59:50 -0700
    • Engine revision fc0e27210c
    • Dart version 2.9.0 (build 2.9.0-19.0.dev 63cf56d925)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/nevercode/Library/Android/sdk
    • Platform android-29, build-tools 29.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.46.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.12.0

[✓] Connected device (5 available)
    • Redmi Note 7       • 6345c469                                 • android-arm64  • Android 9 (API 28)
    • Nevercode’s iPhone • b668e524315069f3db3661ac11ff1f66afafebdb • ios            • iOS 13.5
    • macOS desktop      • macos                                    • darwin-x64     • Mac OS X 10.15.5 19F101
    • Web Server         • web-server                               • web-javascript • Flutter Tools
    • Chrome             • chrome                                   • web-javascript • Google Chrome 83.0.4103.116

• No issues found!
```

</details>

Reproduces also with current stable `1.17.1`, current beta `1.19.0-4.2.pre` and current dev `1.20.0-2.0.pre`. Tried on Android 9 (API 28)

**Comment by aligurelli** [comment]
i faced this problem only page 7 on some device. is there any workaround or fix for this problem?

**Comment by ttencate** [comment]
@jigarpatel17 @VladyslavBondarenko @aligurelli You might want to try if the latest `master` version fixes the problem for you: a fix for the related issue #68182 was recently landed, and I suspect you're seeing the same issue.

**Comment by danagbemava-nc** [comment]
I attempted to reproduce this issue using the code sample in https://github.com/flutter/flutter/issues/60295#issuecomment-651040442 but it did not reproduce for me. I did multiple tries (logs shared below) but in all cases, the times were never apart by more than a few microseconds. 

I'll be closing this as fixed since it does not reproduce. However, if anyone still faces on the latest versions of flutter, kindly file a new issue with updated source code and steps to reproduce so that we can investigate this. 

Thank you

<details>
<summary>updated sample</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(MaterialApp(home: MyApp()));

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final PageController pageController = PageController();
  List colors =
      List.generate(10, (index) => index.isEven ? Colors.red : Colors.blue);

  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: Stack(
        children: [
          PageView.builder(
              itemCount: 10,
              onPageChanged: _handlePageChanged,
              controller: pageController,
              scrollDirection: Axis.vertical,
              physics: const NeverScrollableScrollPhysics(),
              itemBuilder: (context, position) {
                print(
                    'itemBuilder position: $position, time: ${DateTime.now()}');
                return Container(color: colors[position]);
              }),
          Positioned(
            child: ElevatedButton(
              child: const Text('next page'),
              onPressed: () async {
                await pageController.nextPage(
                    duration: const Duration(milliseconds: 100),
                    curve: Curves.bounceIn);
              },
            ),
          )
        ],
      ),
    );
  }

  void _handlePageChanged(int page) {
    print('_handlePageChanged: $page, time: ${DateTime.now()}');
  }
}

```
</details>

<details>
<summary>logs</summary>

```
I/flutter (27398): itemBuilder position: 0, time: 2023-04-20 09:53:31.691052
I/flutter (27398): _handlePageChanged: 1, time: 2023-04-20 09:53:36.683796
I/flutter (27398): itemBuilder position: 1, time: 2023-04-20 09:53:36.689884
I/flutter (27398): itemBuilder position: 2, time: 2023-04-20 09:53:38.563554
I/flutter (27398): _handlePageChanged: 2, time: 2023-04-20 09:53:38.626190
I/flutter (27398): itemBuilder position: 3, time: 2023-04-20 09:53:39.751370
I/flutter (27398): _handlePageChanged: 3, time: 2023-04-20 09:53:39.816609
I/flutter (27398): itemBuilder position: 4, time: 2023-04-20 09:53:40.639059
I/flutter (27398): _handlePageChanged: 4, time: 2023-04-20 09:53:40.692168
I/flutter (27398): itemBuilder position: 5, time: 2023-04-20 09:53:41.404905
I/flutter (27398): _handlePageChanged: 5, time: 2023-04-20 09:53:41.470069
I/flutter (27398): itemBuilder position: 6, time: 2023-04-20 09:53:42.182709
I/flutter (27398): _handlePageChanged: 6, time: 2023-04-20 09:53:42.246765
I/flutter (27398): itemBuilder position: 7, time: 2023-04-20 09:53:43.036761
I/flutter (27398): _handlePageChanged: 7, time: 2023-04-20 09:53:43.102353
I/flutter (27398): itemBuilder position: 8, time: 2023-04-20 09:53:43.915523
I/flutter (27398): _handlePageChanged: 8, time: 2023-04-20 09:53:43.969879
I/flutter (27398): itemBuilder position: 9, time: 2023-04-20 09:53:44.870083
I/flutter (27398): _handlePageChanged: 9, time: 2023-04-20 09:53:44.935029

*==============================================================================*

I/flutter (29484): itemBuilder position: 0, time: 2023-04-20 09:59:58.220270
I/flutter (29484): _handlePageChanged: 1, time: 2023-04-20 10:00:49.107888
I/flutter (29484): itemBuilder position: 1, time: 2023-04-20 10:00:49.119578
I/flutter (29484): itemBuilder position: 2, time: 2023-04-20 10:00:49.160101
I/flutter (29484): _handlePageChanged: 2, time: 2023-04-20 10:00:49.222134
I/flutter (29484): itemBuilder position: 3, time: 2023-04-20 10:00:49.396394
I/flutter (29484): _handlePageChanged: 3, time: 2023-04-20 10:00:49.456257
I/flutter (29484): itemBuilder position: 4, time: 2023-04-20 10:00:49.501522
I/flutter (29484): _handlePageChanged: 4, time: 2023-04-20 10:00:49.570104
I/flutter (29484): itemBuilder position: 5, time: 2023-04-20 10:00:49.623646
I/flutter (29484): _handlePageChanged: 5, time: 2023-04-20 10:00:49.689458
I/flutter (29484): itemBuilder position: 6, time: 2023-04-20 10:00:49.734406
I/flutter (29484): _handlePageChanged: 6, time: 2023-04-20 10:00:49.800130
I/flutter (29484): itemBuilder position: 7, time: 2023-04-20 10:00:49.867544
I/flutter (29484): _handlePageChanged: 7, time: 2023-04-20 10:00:49.933181
I/flutter (29484): itemBuilder position: 8, time: 2023-04-20 10:00:50.025407
I/flutter (29484): _handlePageChanged: 8, time: 2023-04-20 10:00:50.035163
I/flutter (29484): itemBuilder position: 9, time: 2023-04-20 10:00:50.101349
I/flutter (29484): _handlePageChanged: 9, time: 2023-04-20 10:00:50.166286

*==============================================================================*

I/flutter (29484): itemBuilder position: 0, time: 2023-04-20 10:01:43.429395
I/flutter (29484): itemBuilder position: 0, time: 2023-04-20 10:01:43.775604
I/flutter (29484): itemBuilder position: 1, time: 2023-04-20 10:01:51.682500
I/flutter (29484): _handlePageChanged: 1, time: 2023-04-20 10:01:51.735424
I/flutter (29484): itemBuilder position: 2, time: 2023-04-20 10:01:52.337174
I/flutter (29484): _handlePageChanged: 2, time: 2023-04-20 10:01:52.402016
I/flutter (29484): itemBuilder position: 3, time: 2023-04-20 10:01:52.969973
I/flutter (29484): _handlePageChanged: 3, time: 2023-04-20 10:01:53.035034
I/flutter (29484): itemBuilder position: 4, time: 2023-04-20 10:01:53.580953
I/flutter (29484): _handlePageChanged: 4, time: 2023-04-20 10:01:53.634493
I/flutter (29484): itemBuilder position: 5, time: 2023-04-20 10:01:54.279029
I/flutter (29484): _handlePageChanged: 5, time: 2023-04-20 10:01:54.344712
I/flutter (29484): itemBuilder position: 6, time: 2023-04-20 10:01:55.023139
I/flutter (29484): _handlePageChanged: 6, time: 2023-04-20 10:01:55.088179
I/flutter (29484): itemBuilder position: 7, time: 2023-04-20 10:01:55.711692
I/flutter (29484): _handlePageChanged: 7, time: 2023-04-20 10:01:55.776336
I/flutter (29484): itemBuilder position: 8, time: 2023-04-20 10:01:56.378078
I/flutter (29484): _handlePageChanged: 8, time: 2023-04-20 10:01:56.431711
I/flutter (29484): itemBuilder position: 9, time: 2023-04-20 10:01:57.155118
I/flutter (29484): _handlePageChanged: 9, time: 2023-04-20 10:01:57.208879
```
</details>

<details>
<summary>flutter doctor -v</summary>


```
[✓] Flutter (Channel stable, 3.7.12, on macOS 13.3.1 22E261 darwin-arm64, locale en-GB)
    • Flutter version 3.7.12 on channel stable at /Users/nexus/dev/sdks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 4d9e56e694 (2 days ago), 2023-04-17 21:47:46 -0400
    • Engine revision 1a65d409c7
    • Dart version 2.19.6
    • DevTools version 2.20.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.9862592/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode-14.3.0.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.1)
    • Android Studio at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/221.6008.13.2211.9619390/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)

[✓] Android Studio (version 2022.2)
    • Android Studio at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.9862592/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1)
    • IntelliJ at /Users/nexus/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 73.0.4
    • Dart plugin version 231.8109.91

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1)
    • IntelliJ at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/231.8109.175/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.77.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.62.0

[✓] Connected device (4 available)
    • Pixel 7 (mobile)    • 28291FDH2001SA                             • android-arm64  • Android 13 (API 33)
    • M2007J20CG (mobile) • adb-5dd3be00-17AYzd._adb-tls-connect._tcp. • android-arm64  • Android 12 (API 31)
    • macOS (desktop)     • macos                                      • darwin-arm64   • macOS 13.3.1 22E261 darwin-arm64
    • Chrome (web)        • chrome                                     • web-javascript • Google Chrome 112.0.5615.137

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

```
Doctor summary (to see all details, run flutter doctor -v):
[!] Flutter (Channel master, 3.10.0-10.0.pre.17, on macOS 13.3.1 22E261 darwin-arm64, locale en-GB)
    ! Warning: `flutter` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.1)
[✓] Android Studio (version 2022.2)
[✓] IntelliJ IDEA Ultimate Edition (version 2023.1)
[✓] IntelliJ IDEA Ultimate Edition (version 2023.1)
[✓] VS Code (version 1.77.3)
[✓] Connected device (4 available)
[✓] Network resources

! Doctor found issues in 1 category.

```
</details>
